### PR TITLE
python and experiment flow bug fixes

### PR DESCRIPTION
**Pull request by halfway258**
- better NerlnetGetData script
- added nif_join for predict
- comments
- ack and mainserver flow fix
- python flow and send improvement

### CODE REVIEW

File: src_erl/Communication_Layer/http_Nerlserver/src/MainServer/mainGenserver.erl [hunk 1]
```diff
@@ -209,9 +209,9 @@ handle_cast({sourceDone,Body}, State = #main_genserver_state{myName = MyName,sou
 
   case NewCastingList of
     [] -> NextState = State#main_genserver_state{state = idle, sourcesCastingList = NewCastingList,clientsWaitingList = ListOfClients,msgCounter = MsgCounter+1},
-          [{setClientState(clientIdle,ClientName, MyName)}|| ClientName<- ListOfClients],
+          [{setClientState(clientIdle,ClientName, MyName)}|| ClientName<- ListOfClients];
           % gen_server:cast(self(),{clientsIdle}),
-          ack();
+          % ack();
```
Comment (leondavi) [inline src_erl/Communication_Layer/http_Nerlserver/src/MainServer/mainGenserver.erl:214]: @halfway258 
Is it intentional the remove of ack? 
If ack doesn't supposed to be here, you can remove it.

File: src_erl/Communication_Layer/http_Nerlserver/src/nerl_tools.hrl [hunk 2]
```diff
@@ -27,6 +27,7 @@
 %% sorce tmpData name
 -define(TMP_DIRECTORY,"/tmp/nerlnet/").
 -define(TMP_DATA_ADDR, "tmpData.csv").
+-define(DATA_LEN, 1000*1000*1000). % default is 8MB, here set to 1000MB
```
Comment (leondavi) [inline src_erl/Communication_Layer/http_Nerlserver/src/nerl_tools.hrl:30]: @halfway258  
What is this DATA_LEN? what kind of data?

File: src_erl/erlBridge/nerlNIF.erl [hunk 3]
```diff
@@ -182,7 +183,8 @@ nerltensor_conversion({NerlTensor, Type}, ResType) ->
             encode -> Validated = validate_nerltensor_erl(NerlTensor),
                       if
                         Validated -> encode_nif(NerlTensor, BinType);
-                        true -> throw(nerl:string_format("encode failure due to incorrect dimension declaring X*Y*Z not equal to tensor data length! ~p ",[NerlTensor]))
+                        true -> io:format("Bad tensor format!!~n"), {<<>>, BinType}
```
Comment (leondavi) [inline src_erl/erlBridge/nerlNIF.erl:186]: @halfway258 
Change to log error.
Maybe it's better: "Wrong NerlTensor format!"

File: src_py/apiServer/transmitter.py [hunk 4]
```diff
@@ -57,24 +57,19 @@ def updateCSV(self, currentPhase): # currentPhase is either "Training", "Predict
         for root, dirnames, filenames in os.walk(globe.INPUT_DATA_PATH):
             for filename in filenames:
                 if filename == f"{globe.experiment_flow_global.expFlow['CSV path']}_{currentPhase.lower()}.csv":
-                    csvfile = open(os.path.join(root, filename), 'r').readlines()
+                    with open(os.path.join(root, filename), 'r') as f:
```
Comment (leondavi) [inline src_py/apiServer/transmitter.py:60]: Try to avoid naming a variable with a single char.
It can lead to variable shadowing cases.

**Review by leondavi — CHANGES_REQUESTED** [review]

File: src_erl/Communication_Layer/http_Nerlserver/src/MainServer/mainGenserver.erl (hunk 1 above)
Comment (halfway258) [reply src_erl/Communication_Layer/http_Nerlserver/src/MainServer/mainGenserver.erl:214]: removed unnecessary ack

File: src_erl/Communication_Layer/http_Nerlserver/src/nerl_tools.hrl (hunk 2 above)
Comment (halfway258) [reply src_erl/Communication_Layer/http_Nerlserver/src/nerl_tools.hrl:30]: added comment

File: src_erl/erlBridge/nerlNIF.erl (hunk 3 above)
Comment (halfway258) [reply src_erl/erlBridge/nerlNIF.erl:186]: changed message

File: src_py/apiServer/transmitter.py (hunk 4 above)
Comment (halfway258) [reply src_py/apiServer/transmitter.py:60]: changed to fuller name

**Review by leondavi — APPROVED** [review]

**Review by leondavi — APPROVED** [review]

**Pull request merged**